default/python/AI/ShipDesignAI.py

```diff
@@ -1441,6 +1440,66 @@ def _partclass_in_design(self, partclass):
         """
         return any(part.partClass in partclass for part in self.parts)
 
+    def _calculate_weapon_strength(self, weapon_part, ignore_species=False):
+        weapon_name = weapon_part.name
+        base_damage = weapon_part.capacity
+
+        # tech modifiers
+        try:
+            upgrades = AIDependencies.WEAPON_UPGRADE_DICT[weapon_name]
+        except KeyError:
+            if weapon_name not in _raised_warnings:
+                _raised_warnings.add(weapon_name)
+                print_error(("WARNING: Encountered unknown weapon (%s): "
+                             "The AI can play on but its damage estimates may be incorrect leading to worse decision-making. "
+                             "Please update AIDependencies.py and "
+                             "add the weapon with its upgrade techs to WEAPON_UPGRADE_DICT") % weapon_name,
+                            location="ShipDesignAI._calculate_weapon_strength()", trace=True)
+            return base_damage
+        total_tech_bonus = 0
+        for tech, dmg_bonus in upgrades:
+            total_tech_bonus += dmg_bonus if tech_is_complete(tech) else 0
+            # TODO: Error checking if tech is actually a valid tech (tech_is_complete simply returns false)
+
+        # species modifiers
+        if self.species and not ignore_species:
+            weapons_grade = foAI.foAIstate.get_piloting_grades(self.species)
+            species_modifier = AIDependencies.PILOT_DAMAGE_MODIFIER_DICT.get(weapons_grade, {}).get(weapon_name, 0)
+        else:
+            species_modifier = 0
+        return base_damage + total_tech_bonus + species_modifier
+
+    def _calculate_weapon_shots(self, weapon_part, ignore_species=False):
+        weapon_name = weapon_part.name
+        base_shots = weapon_part.secondaryStat
+        if not base_shots:
+            print "WARNING: Queried weapon %s for number of shots but didn't return any." % weapon_name
+            base_shots = 1
+
+        # species modifier
+        if self.species and not ignore_species:
+            weapons_grade, _, _ = foAI.foAIstate.get_piloting_grades(self.species)
+            species_modifier = AIDependencies.PILOT_ROF_MODIFIER_DICT.get(weapons_grade, {}).get(weapon_name, 0)
+        else:
+            species_modifier = 0
+
+        # tech modifier
+        tech_modifier = 0  # none implemented yet
+        return base_shots + species_modifier + tech_modifier
+
+    def _calculate_hangar_damage(self, hangar_part, ignore_species=False):
```

Comment:
It's not strictly the same. In this function, the damage of a single part should be calculated.

How that damage of the hangar part then translates into final fighter damage should be covered in another part of the code, IMO as that depends on another implementation detail on the C++ side (which may still be subject to change in the future considering the feature is far from established).